src/librustc_mir/util/elaborate_drops.rs

```diff
@@ -858,32 +860,75 @@ impl<'l, 'b, 'tcx, D> DropCtxt<'l, 'b, 'tcx, D>
 
     fn box_free_block<'a>(
         &mut self,
-        ty: Ty<'tcx>,
+        adt: &'tcx ty::AdtDef,
+        substs: &'tcx Substs<'tcx>,
         target: BasicBlock,
         unwind: Unwind,
     ) -> BasicBlock {
-        let block = self.unelaborated_free_block(ty, target, unwind);
+        let block = self.unelaborated_free_block(adt, substs, target, unwind);
         self.drop_flag_test_block(block, target, unwind)
     }
 
     fn unelaborated_free_block<'a>(
         &mut self,
-        ty: Ty<'tcx>,
+        adt: &'tcx ty::AdtDef,
+        substs: &'tcx Substs<'tcx>,
         target: BasicBlock,
         unwind: Unwind
     ) -> BasicBlock {
         let tcx = self.tcx();
         let unit_temp = Place::Local(self.new_temp(tcx.mk_nil()));
         let free_func = tcx.require_lang_item(lang_items::BoxFreeFnLangItem);
-        let substs = tcx.mk_substs(iter::once(Kind::from(ty)));
+        let free_sig = tcx.fn_sig(free_func).skip_binder().clone();
+        let free_inputs = free_sig.inputs();
+        // If the box_free function takes a *mut T, transform the Box into
+        // such a pointer before calling box_free. Otherwise, pass it all
+        // the fields in the Box as individual arguments.
```

Comment:
I think supporting the `*mut T` form adds unnecessary clutter to the compiler.

Reply:
`#[cfg(stage0)]` `box_free` is *tiny* compared to this, since all it needs to do is call the real implementation, which it can do in much less code. (See my patch for more details)

Reply:
The other problem is that a `#[cfg(stage0)]` box_free with the old signature can't be compiled by the new code, which means if, say, version 1.27 has the new code, you can't bootstrap 1.27.1 (if there ends up being one) with 1.27, which, typically, linux distros would be doing.

Reply:
Ehm, we don't support that - or rather, you set a flag in config.toml that turns off stage0. cc @alexcrichton

Reply:
(we pretty much always use `#[cfg(stage0)]` like this)